src/ireiat/data_pipeline/assets/demand/faf5_tons.py

```diff
@@ -11,54 +11,155 @@
 
-    # Initialize a list to store DataFrames for each mode
-    demand_dataframes = []
+    # Split into known modes (truck, rail, water) and unknown modes
+    known_modes_pdf = faf_containerizable_pdf[
+        faf_containerizable_pdf["dms_mode"].isin([FAFMode.TRUCK, FAFMode.RAIL, FAFMode.WATER])
+    ]
+    unknown_modes_pdf = faf_containerizable_pdf[
+        faf_containerizable_pdf["dms_mode"] == FAFMode.OTHER_AND_UNKNOWN
+    ]
+
+    # Aggregate by origin/destination for known modes
+    known_agg_pdf = known_modes_pdf.groupby(["dms_orig", "dms_dest", "dms_mode"], as_index=False)[
+        [FAF_TONS_TARGET_FIELD]
+    ].sum()
+
+    # Pivot by origin/destination to calculate percentages for each mode
+    mode_pivot_pdf = known_agg_pdf.pivot_table(
+        index=["dms_orig", "dms_dest"],
+        columns="dms_mode",
+        values=FAF_TONS_TARGET_FIELD,
+        fill_value=0,
+    )
+
+    # Add total tons and calculate percentages
+    mode_pivot_pdf["total_tons"] = mode_pivot_pdf.sum(axis=1)
+    mode_pivot_pdf["truck_pct"] = mode_pivot_pdf[FAFMode.TRUCK] / mode_pivot_pdf["total_tons"]
+    mode_pivot_pdf["rail_pct"] = mode_pivot_pdf[FAFMode.RAIL] / mode_pivot_pdf["total_tons"]
+    mode_pivot_pdf["water_pct"] = mode_pivot_pdf[FAFMode.WATER] / mode_pivot_pdf["total_tons"]
+
+    # Filter unknown origin/destination pairs that don't exist in known modes
+    valid_orig_dest_pairs = known_agg_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+
+    # Calculate unique origin-destination pairs in unknown modes before filtering
+    original_unknown_pairs = unknown_modes_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+
+    # Merge unknown modes with the known mode proportions, filter invalid pairs
+    unknown_modes_pdf = unknown_modes_pdf.merge(
+        mode_pivot_pdf[["truck_pct", "rail_pct", "water_pct"]],
+        on=["dms_orig", "dms_dest"],
+        how="left",
+    )
 
-    # Loop through each mode and generate the corresponding demand DataFrame
-    for mode in FAFDemandByMode:
-        mode_str = mode.name  # e.g., "TRUCK", "RAIL"
+    # Filter unknown origin/destination pairs that don't exist in known modes by merging with valid pairs
+    unknown_modes_pdf = unknown_modes_pdf.merge(
+        valid_orig_dest_pairs, on=["dms_orig", "dms_dest"], how="inner"
+    )
 
-        is_by_mode = faf_containerizable_pdf["dms_mode"] == FAFMode[mode_str]
-        faf_mode_pdf = faf_containerizable_pdf.loc[is_by_mode]
+    # Calculate the dropped pairs
+    remaining_unknown_pairs = unknown_modes_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+    dropped_pairs_count = len(original_unknown_pairs) - len(remaining_unknown_pairs)
 
-        # Group by origin and destination to calculate total tons
-        total_tons_od_pdf = faf_mode_pdf.groupby(["dms_orig", "dms_dest"], as_index=False)[
-            [FAF_TONS_TARGET_FIELD]
-        ].sum()
+    # Log a warning about the dropped origin-destination pairs
+    if dropped_pairs_count > 0:
```

Comment:
I think this is fine for now...but it does make me a bit nervous to just drop stuff.

Do you have any idea of the magnitude of tons that get dropped? And do you mind putting a TODO here to "figure out a different way to handle FAF zone pairs that don't appear in other modes".

A crude solution might be some kind of regression for each mode based on the origin zone (categorical variable), destination zone (categorical variable), and maybe the haversine distance between them...

Reply:
Added a TODO for us to revisit in the future